R/build_lm.R

```diff
@@ -130,3 +130,157 @@ build_lm <- function(data, formula, ..., keep.source = TRUE, augment = FALSE, gr
   }
   ret
 }
+
+#' builds lm model quickly for analytics view.
+#' @export
+build_lm.fast <- function(df,
+                    target,
+                    ...,
+                    max_nrow = 200000,
+                    predictor_n = 10
+                    ){
+  # TODO: cleanup code only aplicable to randomForest. this func was started from copy of calc_feature_imp, and still adjusting for lm. 
+
+  # this seems to be the new way of NSE column selection evaluation
+  # ref: https://github.com/tidyverse/tidyr/blob/3b0f946d507f53afb86ea625149bbee3a00c83f6/R/spread.R
+  target_col <- dplyr::select_var(names(df), !! rlang::enquo(target))
+  # this evaluates select arguments like starts_with
+  selected_cols <- dplyr::select_vars(names(df), !!! rlang::quos(...))
+
+  grouped_cols <- grouped_by(df)
+
+  # remove grouped col or target col
+  selected_cols <- setdiff(selected_cols, c(grouped_cols, target_col))
+
+  if (any(c(target_col, selected_cols) %in% grouped_cols)) {
+    stop("grouping column is used as variable columns")
+  }
+
+  # remove NA because it's not permitted for randomForest. TODO is this good for lm too?
+  df <- df %>%
+    dplyr::filter(!is.na(!!target_col))
+
+  # cols will be filtered to remove invalid columns
+  cols <- selected_cols
+
+  for (col in selected_cols) {
+    if(all(is.na(df[[col]]))){
+      # remove columns if they are all NA
+      cols <- setdiff(cols, col)
+    }
+  }
+
+  # randomForest fails if columns are not clean. TODO is this needed?
+  #clean_df <- janitor::clean_names(df)
+  clean_df <- df # turn off clean_names for lm
+  # this mapping will be used to restore column names
+  name_map <- colnames(clean_df)
+  names(name_map) <- colnames(df)
+
+  # clean_names changes column names
+  # without chaning grouping column name
+  # information in the data frame
+  # and it causes an error,
+  # so the value of grouping columns
+  # should be still the names of grouping columns
+  name_map[grouped_cols] <- grouped_cols
+  colnames(clean_df) <- name_map
+
+  clean_target_col <- name_map[target_col]
+  clean_cols <- name_map[cols]
+
+  each_func <- function(df) {
+    tryCatch({
+      # sample the data because randomForest takes long time
+      # if data size is too large
+      if (nrow(df) > max_nrow) {
+        df <- df %>%
+          dplyr::sample_n(max_nrow)
```

Comment:
You may want to set.seed to make the output consistent.

Reply:
added seed parameter.